### PR TITLE
Fix small discovery radius under normal conditions or when all multipliers are disabled

### PR DESCRIPTION
**Pull request by richteer**
Hello, thanks for this updated serversync'd version!

I noticed when playing that the visibility radius appeared to be very small, typically the clamped minimum of `35.0` units, even in fairly "normal" conditions -- reasonable light, sea level, clear skies, meadows. Furthermore, disabling each of the different multipliers (setting to `0.0`) also yielded a discovery radius of that minimum `35.0`.

After some poking around, it appears that the multipliers for each of the conditions are summed, but without a base value of `1.0f`, these "normal" or "disabled" conditions are summed to a multiplier of `0.0f`, which is forcing the clamped discovery radius. Inspecting Pathfinder's source shows that there is a [similar base multiplier of `1.0f`](https://github.com/CrystalFerrai/ValheimMods/blob/master/Pathfinder/PathfinderPlugin.cs#L235).

Note this will dramatically increase the discovery radius, though I'm currently playing with a patched version and the current defaults seem to be a fair discovery range. This base multiplier value *could* be set as a config value, though to be honest I'm not sure what the benefits of setting anything other than `1.0f` would be -- at that point the base radius value should be changed, not a base radius multiplier.

### CODE REVIEW

**Comment by FelixReuthlinger** [comment]
Hey, thanks for the PR. I tried to understand, what you want to achieve with this change, but I am not sure if I got it right. If I add 1.0f to the multiplier or just leave sum to 0.0f does in the end not really change much in the result, since clamping the result will always set the minimum to 35. I guessed that a minimum visible discovery of 35 meters is ok from a gameplay perspective. Instead of the change to add 1.0f I could also easily let you configure the min clamp value instead, would that make more sense?

**Comment by FelixReuthlinger** [comment]
I added the min/max config option and released v1.1.0

**Comment by richteer** [comment]
Ah sorry, let me try to reword and elaborate my rambly mess. I started writing this before the recently pushed changes, so bear with any inconsistencies.

---
## The Problem

The issue comes from the expectation of how the mod works based on the configuration settings and their respective help text. The expectation from setting `LandExploreRadius` to `150.0f` with generally "normal conditions"[1], the exploration radius would be `150.0f`. Furthermore, a user would expect that changes to the environment (and thus the multipliers) would then grow or shrink this radius depending on how severe the conditions are. 

The problem comes in under these "normal conditions", all the multipliers would be `0.0f`, as in "not contributing to the increase or decrease of `LandExploreRadius`". By taking the result of the summed multipliers of `0.0f`, the value of `LandExploreRadius` is irrelevant, since "normal conditions" will always translate to the min clamp. This also means that at sea, `SeaExploreRadius` will also be irrelevant since that will also translate to min clamp, leading there to be no meaningful separation from land vs sea exploration.

By adding `1.0f` to the summed multipliers, this is equivalent to `baseRadius + (baseRadius * radiusTotalMultiplier)`, where most importantly the product of the values in parentheses will be `0.0f` under "normal conditions", therefore leaving the final sum at the `baseRadius`, or `LandExploreRadius` on land and `SeaExploreRadius` at sea.

[1] "Normal conditions" implying average light, no weather, sea-level, clear biome like the meadows


## Current Implementation

To go a little more into what the mod is *currently* doing, the current `ExploreRadius` config values essentially set a "maximum explore radius", since under normal conditions the post-multiplier value under normal conditions will always be min clamp, and the explore radius can only be raised above min clamp if the multipliers sum to positive (technically, the sum of the multipliers needs to be above `baseRadius / min clamp`, as a small positive multiplier like 0.001 may still result in a positive radius below min clamp). However, it isn't exactly even a maximum, since the sum of the multipliers could exceed `1.0f` under very favorable conditions. It is closer to an "exploration scale factor", but how the scale factor is applied is not clear to a user.

From my understanding, the four modifiers have the following effects on the radius:
- lighting can increase or decrease
- weather can only decrease
- altitude can only increase
- biome can only decrease

Weather and biome both only decrease the exploration radius, therefore if we are clamped to the minimum under "normal conditions", the exploration radius will not shrink when entering a forest, or when it starts to rain. Similarly, if always at the min clamp value, a particularly dominant negative multiplier will prevent any other positive multiplier from showing any effect until that dominant negative multiplier has been overcome. For example, night time with a lighting multiplier of `-1.0f`, going up in elevation would not cause a linear increase in exploration radius as expected, it would stay at min clamp until halfway up a mountain, where then it will start increasing linearly.

### Scenario: Disabling All Multipliers

The other scenario is where a user does not want to use any of the multipliers, and only use this mod to adjust the land and sea exploration radius. The expectation would be that by setting all the multipliers to `0.0f`, light, weather, biome and altitude would not alter the radius at all. However the result is that both the `LandExploreRadius` and `SeaExploreRadius` values are entirely ignored, and only the min clamp value is used.


## Side Note

One side effect however, is that the amount the radius grows or shrinks based on the multipliers will also scale with the respective setting for `LandExploreRadius`/`SeaExploreRadius`. For example, if `baseRadius = 100.0f` and `radiusTotalMultiplier = 0.10f`, then the resulting exploration radius is `110.0f`, an increase of `10.0f`. If `baseRadius = 200.0f` instead, then the result with the same multiplier would be `220.0f`, an increase of `20.0f`. This would lead to slightly more dramatic changes in radius with a higher `ExploreRadius` setting, though that can be counteracted by reduced the contribution of each multiplier slightly.

An alternative would be to introduce a configuration setting for `radiusScaleFactor`, and do the final radius calculation something like `baseRadius + (radiusScaleFactor * radiusTotalMultiplier)`. In this case, the radius will always grow or shrink a consistent amount regardless of `baseRadius`. This may be out of scope for this PR though and would warrant its own discussion.


---
TL;DR/Notes:
 - Always sitting at min clamp means negative multipliers never shrink the radius, only prevent positive multipliers from growing the radius until positive multipliers are large enough (which is almost never, in reality)
 - Adjusting min/max clamp should be a nice way to restrict how the radius grows/shrinks, but ideally should only be hit in extreme scenarios.

**Comment by FelixReuthlinger** [comment]
Really respect that you put that much effort in explaining your thoughts :)
The fact that someone would want to deactivate all of the four multipliers at once (set all to 0.0f) from my perspective would be rather a corner case. Since if you don't want to use the mods features, why use the mod at all, you can also easily go without the mod if you prefer a discovery radius that does not vary by conditions. 
But if I get it right this is not your point, is it? There might be a quite small chance that the different calculated multipliers (having the configured values applied together with the current game conditions the player is in) sum up to 0.0f:
- lighting yields only positive values (if I remeber correctly)
- weather only creates negative values
- altitude only create positive values
- biome (and dependent on if inside forest) can either create positive or negative values

So, there is a small chance that this will sum up to 0.0f, which then will limit discovery to the min radius. But usually these conditions can change that quickly that you would not necessarily recognize it. Or does it stay the same for you for a very long time?
And even if this is the case that it is 0.0f for some time, if I would just add 1.0f, there is still the chance that this value gets to 0.0f, we just moved the window for matching conditions by this a tiny bit, right? If instead of moving it, which does not solve it, I would clamp the multipliers to a minimum of 1.0f, the effect would also yield not desired effects. Since then we would never reach the minimum discovery distance, since using e.g. 150 base radius * 1 = 150 -> clamp => still 150. With this bad conditions can never limit your visibility, but from my perspective this is what I wanted to also achieve, since running through fog I sometimes can spot objects on the minimap faster than in my first person view, which is undesired from my point of view.

**Comment by richteer** [comment]
> Really respect that you put that much effort in explaining your thoughts :)

I've found that sometimes the smallest code changes need the largest explanations :)

> The fact that someone would want to deactivate all of the four multipliers at once (set all to 0.0f) from my perspective would be rather a corner case. Since if you don't want to use the mods features, why use the mod at all, you can also easily go without the mod if you prefer a discovery radius that does not vary by conditions.

You're correct in that this is not my point, but there is one value in using the mod without modifiers: easy configuration of land vs sea exploration radii -- that's actually how I found this mod in the first place, and the dynamic nature of it was bonus points. I can't say I looked exhaustively, but I did not find another mod that had that kind of configuration. Of course, that is indeed, not the point here.

> There might be a quite small chance that the different calculated multipliers (having the configured values applied together with the current game conditions the player is in) sum up to 0.0f:
> 
>     * lighting yields only positive values (if I remeber correctly)
> 

This appears to be sometimes slightly negative at night (`-0.1f` with the default config), though generally positive.

> 
> So, there is a small chance that this will sum up to 0.0f, which then will limit discovery to the min radius. But usually these conditions can change that quickly that you would not necessarily recognize it. Or does it stay the same for you for a very long time?

With default configs at roughly sea level in the meadows at noon with no weather, I'm seeing the lighting multiplier at `0.4f`, and the biome multiplier at `-0.4f`, resulting in a sum multiplier of `0.0f`, therefore getting the min radius.

> And even if this is the case that it is 0.0f for some time, if I would just add 1.0f, there is still the chance that this value gets to 0.0f, we just moved the window for matching conditions by this a tiny bit, right? If instead of moving it, which does not solve it, I would clamp the multipliers to a minimum of 1.0f, the effect would also yield not desired effects. 

This is where I think the core of it is -- the multipliers should adjust the base radius up or down based on conditions, however in the current state, the exploration radius is entirely dependent on the multipliers. No light or altitude = no exploration radius. In this case, `1.0f` is the *base* multiplier, and the other multipliers are either increasing or decreasing from the base.

> Since then we would never reach the minimum discovery distance, since using e.g. 150 base radius * 1 = 150 -> clamp => still 150. With this bad conditions can never limit your visibility, but from my perspective this is what I wanted to also achieve, 

Clamping the multiplier to a minimum `1.0f` is definitely not the intended change, biome and weather with the current defaults can be more than enough to bring the multiplier back down to below the radius clamping threshold. With a base radius of `150.f`, you need to have a summed multiplier higher than `0.233f` to be above the `35.0f` default min clamp. With my example before of sea-level meadows at noon and a base multiplier of `1.0f`, the sum modifier would be `1.0f + 0.4f - 0.4f + 0.0f + 0.0f` -> `1.0f`, yielding an exploration radius of `150.0f`. If I were to go to the nearby black forest, the change would be like so: `1.0f + 0.4f - 0.6f + 0.0f + 0.0f` -> `-0.8f`, yielding an exploration radius of `120.0f`.

Without a base multiplier of `1.0f`, both the meadows and black forest would be at the same radius of `35.0f`, since the total multiplier would be `0.0f` in the meadows, and `-0.2f` in the black forest, yielding the min clamp radius in both cases. The change in biome would not have changed the exploration radius at all.

As a side note: without the base multiplier, changing the base radius would also have no effect on the exploration radius either, the base radius only comes in to play once the summed multipliers exceed a particular value, which then the radius will start to grow.

> since running through fog I sometimes can spot objects on the minimap faster than in my first person view, which is undesired from my point of view.

This should still be able to be tweaked in the config -- since adding a base multiplier will generally increase exploration radius, the base radius should probably be reduced.

**Comment by FelixReuthlinger** [comment]
>the sum modifier would be 1.0f + 0.4f - 0.4f + 0.0f + 0.0f -> 1.0f, yielding an exploration radius of 150.0f

That is actually what is not intended, and that is also why I still do not see a point in just adding +1.0f, since then even in conditions not being optimal you will get a discovery radius that is far beyond natural sight with those conditions. I did really evaluate the settings that are used as default by running around the map with all different conditions and always trying to evaluate if the exploration radius is similar to the visual line of sight for spotting my surrounding on the map. You can try to evaluate yourself, of course and adjust settings as you like using the mod, but I still do not see a point in boosting the exploration beyond natural line of sight.
